src/charm.py

```diff
@@ -18,3 +28,65 @@ def __init__(self, *args):
             args: Arguments passed to the CharmBase parent constructor.
         """
         super().__init__(*args)
+        self._charm_state = None
+        self._saml_integrator = None
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.saml_relation_created, self._on_saml_relation_created)
+
+    def _on_upgrade_charm(self, _) -> None:
+        """Install needed apt packages."""
+        self.unit.status = ops.MaintenanceStatus("Installing packages")
```

Comment:
Is `MaintainanceStatus` not removed by setting the status to active after the apt operations are finished because we expect the `config-changed` event to fire afterwards as the lifecycle guarantees? 
Would setting the state to active explicitly be an option?

Reply:
Exactly. The config changed will be fired afterwwards. We could do that, but I don't see the need